pika/channel.py

```diff
@@ -944,32 +1017,81 @@ def _on_cancelok(self, method_frame):
     def _on_close(self, method_frame):
         """Handle the case where our channel has been closed for us
 
-        :param pika.frame.Method method_frame: The close frame
+        :param pika.frame.Method method_frame: Method frame with Channel.Close
+            method
 
         """
-        LOGGER.info('%s', method_frame)
-        LOGGER.warning('Received remote Channel.Close (%s): %r on channel %s',
+        LOGGER.warning('Received remote Channel.Close (%s): %r on %s',
                        method_frame.method.reply_code,
                        method_frame.method.reply_text,
                        self)
-        if self.connection.is_open:
-            self._send_method(spec.Channel.CloseOk())
-        self._set_state(self.CLOSED)
-        self.callbacks.process(self.channel_number, '_on_channel_close', self,
-                               self, method_frame.method.reply_code,
-                               method_frame.method.reply_text)
-        self._cleanup()
+
+        # AMQP 0.9.1 requires CloseOk response to Channel.Close; Note, we should
+        # not be called when connection is closed
+        self._send_method(spec.Channel.CloseOk())
+
+        # This overrides the prior value if connection was in CLOSING state
+        self._closing_code_and_text = (method_frame.method.reply_code,
+                                       method_frame.method.reply_text)
+
+        if self.is_closing:
+            # Since we already sent Channel.Close, we need to wait for CloseOk
+            # before cleaning up to avoid a race condition whereby our channel
+            # number might get reused before our CloseOk arrives
+            pass
+        else:
+            self._set_state(self.CLOSED)
+            try:
+                self.callbacks.process(self.channel_number, '_on_channel_close',
+                                       self, self,
+                                       method_frame.method.reply_code,
+                                       method_frame.method.reply_text)
+            finally:
+                self._cleanup()
+
+    def _on_close_meta(self, reply_code, reply_text):
```

Comment:
REVIEWME new dedicated function for use by `Connection` class when connection is lost abruptly and channels remain that need to be cleaned up.